File: 0140-word-break-ii/0140-word-break-ii.cpp
class Solution {
public:
     vector<string> ans;
    bool solve(string s,vector<string>& words,string temp)
    {
        if(s.size() == 0)
        {
            ans.push_back(temp);
            return true;
        }
        bool flag = 0;
        for(int i = 1; i <= s.size(); i++)
        {
            string str = s.substr(0,i);
            bool f = 0;
            for(int j = 0; j < words.size(); j++)
            {
                if(str == words[j])
                {
                    f = 1;
                    break;
                }
            }
            if(f and solve(s.substr(i),words,temp + str + ' ')) flag = 1;
        }
        return flag;
    }
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        solve(s,wordDict,"");
        for(int i = 0; i < ans.size(); i++)
        {
            ans[i].pop_back();
        }
        return ans;
    }
};

File: 0140-word-break-ii/0140-word-break-ii.java
class Solution {
     public List<String> wordBreak(String s, List<String> wordDict) {
    List<String> result = new ArrayList<>();
    HashSet<String> set = new HashSet<>(wordDict);
    helper(s,set, 0, 0, "", List.of(),result);
    return result;
  }

  public void helper(String s, Set<String> set, int start, int end, String word, List<String> words, List<String> res){
    System.out.println(String.format("start: %d y end: %d", start, end));
    if(word.equals(s)){
      res.add(String.join(" ", words));
    }
    else if(start >= s.length() || end > s.length()) return;
    else{
      String prefix = s.substring(start, end);
      if(set.contains(prefix)){
        ArrayList<String> temp = new ArrayList<>(words);
        temp.add(prefix);
        helper(s, set, end, end, word + prefix, temp, res);
        helper(s, set, start, end + 1, word, new ArrayList<>(words),res);
      }else{
        helper(s, set, start, end + 1, word, new ArrayList<>(words), res);
      }
    }
  }
}